test/account_api_test.rb

```diff
@@ -1,81 +1,407 @@
+        .upon_receiving("a sign-in request")
+        .with(
+          method: :get,
+          path: "/api/oauth2/sign-in",
+          headers: GdsApi::JsonClient.default_request_headers,
+        )
+        .will_respond_with(
+          status: 200,
+          headers: { "Content-Type" => "application/json; charset=utf-8" },
+          body: {
+            auth_uri: Pact.like("http://authentication-provider/some/oauth/url"),
+            state: Pact.like("value-to-use-for-csrf-prevention"),
+          },
+        )
+    end
+
+    it "responds with 200 OK, an authentication URI, and a state for CSRF protection" do
+      response = api_client.get_sign_in_url
+      assert response["auth_uri"].present?
+      assert response["state"].present?
+      assert_equal 200, response.code
+    end
   end
 
-  it "gives a session ID if the auth response validates" do
-    stub_account_api_validates_auth_response(code: "foo", state: "bar")
-    assert(!api_client.validate_auth_response(code: "foo", state: "bar")["govuk_account_session"].nil?)
+  describe "getting a sign-in URL, with a redirect_path pointing to another domain" do
+    let(:redirect_path) { "//protocol-relative-redirect" }
+
+    before do
+      account_api
+        .given("no particular state")
+        .upon_receiving("a sign-in request with an invalid redirect_path")
+        .with(
+          method: :get,
+          path: "/api/oauth2/sign-in",
+          headers: GdsApi::JsonClient.default_request_headers,
+          query: { redirect_path: Pact.like(redirect_path) },
+        )
+        .will_respond_with(
+          status: 422,
+          headers: { "Content-Type" => "application/json; charset=UTF-8" },
+        )
+    end
+
+    it "responds with 422 Unprocessable Entity" do
+      assert_raises GdsApi::HTTPUnprocessableEntity do
+        api_client.get_sign_in_url(redirect_path: redirect_path)
+      end
+    end
   end
 
-  it "throws a 401 if the auth response does not validate" do
-    stub_account_api_rejects_auth_response(code: "foo", state: "bar")
+  describe "validating an OAuth response (success)" do
+    let(:given) { "there is a valid OAuth response" }
+    let(:redirect_path) { nil }
 
-    assert_raises GdsApi::HTTPUnauthorized do
-      api_client.validate_auth_response(code: "foo", state: "bar")
+    before do
+      account_api
+        .given(given)
+        .upon_receiving("a validation request")
+        .with(
+          method: :post,
+          path: "/api/oauth2/callback",
+          body: {
+            code: Pact.like("code"),
+            state: Pact.like("state"),
+          },
+          headers: GdsApi::JsonClient.default_request_with_json_body_headers,
+        )
+        .will_respond_with(
+          status: 200,
+          headers: { "Content-Type" => "application/json; charset=utf-8" },
+          body: {
+            govuk_account_session: Pact.like("user-session-id"),
+            redirect_path: Pact.like(redirect_path),
```

Comment:
I found it a bit confusing that `redirect_path` is nil. What does Pact.like(nil) allow to pass?

Reply:
Ah, this can go away now.  It was useful when I had the test for the error case, so I didn't have to write two almost-identical before blocks, but now that's gone.